File: app/src/main/java/com/aricneto/twistytimer/layout/RippleBackgroundView.java
package com.aricneto.twistytimer.layout;

import android.animation.Animator;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.widget.FrameLayout;

import com.aricneto.twistify.R;

import java.util.ArrayList;

public class RippleBackgroundView extends FrameLayout {

    private int rippleColor;
    private float rippleStrokeWidth;
    private float rippleRadius;
    private int rippleDurationTime;
    private int rippleAmount;
    private int rippleDelay;
    private float rippleScale;
    private int rippleType;

    private Paint paint;
    private boolean animationRunning = false;
    private AnimatorSet animatorSet;
    private ArrayList<RippleView> rippleViewList = new ArrayList<>();

    public RippleBackgroundView(Context context) {
        super(context);
    }

    public RippleBackgroundView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    public RippleBackgroundView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    private void init(final Context context, final AttributeSet attrs) {
        if (isInEditMode()) return;

        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.RippleBackgroundView);
        rippleColor = typedArray.getColor(R.styleable.RippleBackgroundView_rb_color, getResources().getColor(R.color.md_blue_A700));
        rippleStrokeWidth = typedArray.getDimension(R.styleable.RippleBackgroundView_rb_strokeWidth, getResources().getDimension(R.dimen.rippleStrokeWidth));
        rippleRadius = typedArray.getDimension(R.styleable.RippleBackgroundView_rb_radius, getResources().getDimension(R.dimen.rippleRadius));
        rippleDurationTime = typedArray.getInt(R.styleable.RippleBackgroundView_rb_duration, 3000);
        rippleAmount = typedArray.getInt(R.styleable.RippleBackgroundView_rb_rippleAmount, 6);
        rippleScale = typedArray.getFloat(R.styleable.RippleBackgroundView_rb_scale, 6.0f);
        rippleType = typedArray.getInt(R.styleable.RippleBackgroundView_rb_type, 0);
        typedArray.recycle();

        rippleDelay = rippleDurationTime / rippleAmount;

        paint = new Paint();
        paint.setAntiAlias(true);
        if (rippleType == 0) {
            rippleStrokeWidth = 0;
            paint.setStyle(Paint.Style.FILL);
        } else {
            paint.setStyle(Paint.Style.STROKE);
        }
        paint.setColor(rippleColor);

        LayoutParams rippleParams = new LayoutParams((int) (2 * (rippleRadius + rippleStrokeWidth)), (int) (2 * (rippleRadius + rippleStrokeWidth)));
        rippleParams.gravity = android.view.Gravity.CENTER;

        animatorSet = new AnimatorSet();
        animatorSet.setInterpolator(new AccelerateDecelerateInterpolator());
        ArrayList<Animator> animators = new ArrayList<>();

        for (int i = 0; i < rippleAmount; i++) {
            RippleView rippleView = new RippleView(getContext());
            addView(rippleView, rippleParams);
            rippleViewList.add(rippleView);
            final ObjectAnimator scaleXAnimator = ObjectAnimator.ofFloat(rippleView, "ScaleX", 1.0f, rippleScale);
            scaleXAnimator.setRepeatCount(ObjectAnimator.INFINITE);
            scaleXAnimator.setRepeatMode(ObjectAnimator.RESTART);
            scaleXAnimator.setStartDelay(i * rippleDelay);
            scaleXAnimator.setDuration(rippleDurationTime);
            animators.add(scaleXAnimator);

            final ObjectAnimator scaleYAnimator = ObjectAnimator.ofFloat(rippleView, "ScaleY", 1.0f, rippleScale);
            scaleYAnimator.setRepeatCount(ObjectAnimator.INFINITE);
            scaleYAnimator.setRepeatMode(ObjectAnimator.RESTART);
            scaleYAnimator.setStartDelay(i * rippleDelay);
            scaleYAnimator.setDuration(rippleDurationTime);
            animators.add(scaleYAnimator);

            final ObjectAnimator alphaAnimator = ObjectAnimator.ofFloat(rippleView, "Alpha", 1.0f, 0f);
            alphaAnimator.setRepeatCount(ObjectAnimator.INFINITE);
            alphaAnimator.setRepeatMode(ObjectAnimator.RESTART);
            alphaAnimator.setStartDelay(i * rippleDelay);
            alphaAnimator.setDuration(rippleDurationTime);
            animators.add(alphaAnimator);
        }

        animatorSet.playTogether(animators);
    }

    private class RippleView extends View {

        public RippleView(Context context) {
            super(context);
            this.setVisibility(View.INVISIBLE);
        }

        @Override
        protected void onDraw(Canvas canvas) {
            int radius = (Math.min(getWidth(), getHeight())) / 2;
            canvas.drawCircle(radius, radius, radius - rippleStrokeWidth, paint);
        }
    }

    public void startRippleAnimation() {
        if (!isRippleAnimationRunning()) {
            for (RippleView rippleView : rippleViewList) {
                rippleView.setVisibility(VISIBLE);
            }
            animatorSet.start();
            animationRunning = true;
        }
    }

    public void stopRippleAnimation() {
        if (isRippleAnimationRunning()) {
            animatorSet.end();
            animationRunning = false;
            for (RippleView rippleView : rippleViewList) {
                rippleView.setVisibility(INVISIBLE);
            }
        }
    }

    public boolean isRippleAnimationRunning() {
        return animationRunning;
    }
}

File: app/src/main/java/com/aricneto/twistytimer/layout/RippleBackgroundView.kt
package com.aricneto.twistytimer.layout

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.animation.AccelerateDecelerateInterpolator
import android.view.animation.LinearInterpolator
import android.widget.FrameLayout
import androidx.core.content.withStyledAttributes
import com.aricneto.twistify.R
import com.google.android.material.color.MaterialColors

class RippleBackgroundView : FrameLayout {
    private var rippleColor = 0
    private var rippleStrokeWidth = 0f
    private var rippleRadius = 0f
    private var rippleDurationTime = 0
    private var rippleAmount = 0
    private var rippleScale = 0f
    private var rippleType = 0

    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val interpolator = AccelerateDecelerateInterpolator()
    
    var isRippleAnimationRunning: Boolean = false
        private set
    private var animator: ValueAnimator? = null

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(context, attrs)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        init(context, attrs)
    }

    private fun init(context: Context, attrs: AttributeSet?) {
        if (isInEditMode) return
        setWillNotDraw(false)

        context.withStyledAttributes(attrs, R.styleable.RippleBackgroundView) {
            rippleColor = getColor(
                R.styleable.RippleBackgroundView_rb_color,
                MaterialColors.getColor(context, R.attr.colorTertiary, Color.BLUE)
            )
            rippleStrokeWidth = getDimension(
                R.styleable.RippleBackgroundView_rb_strokeWidth, resources.getDimension(
                    R.dimen.rippleStrokeWidth
                )
            )
            rippleRadius = getDimension(
                R.styleable.RippleBackgroundView_rb_radius, resources.getDimension(
                    R.dimen.rippleRadius
                )
            )
            rippleDurationTime = getInt(R.styleable.RippleBackgroundView_rb_duration, 3000)
            rippleAmount = getInt(R.styleable.RippleBackgroundView_rb_rippleAmount, 6)
            rippleScale = getFloat(R.styleable.RippleBackgroundView_rb_scale, 6.0f)
            rippleType = getInt(R.styleable.RippleBackgroundView_rb_type, 0)
        }

        paint.color = rippleColor
        paint.strokeWidth = rippleStrokeWidth
        paint.style = if (rippleType == 0) Paint.Style.FILL else Paint.Style.STROKE

        animator = ValueAnimator.ofFloat(0f, 1.0f).apply {
            duration = rippleDurationTime.toLong()
            repeatCount = ValueAnimator.INFINITE
            repeatMode = ValueAnimator.RESTART
            interpolator = LinearInterpolator()
            addUpdateListener { postInvalidateOnAnimation() }
        }
    }

    override fun onDraw(canvas: Canvas) {
        if (!isRippleAnimationRunning) return

        val centerX = width / 2f
        val centerY = height / 2f
        val progress = animator?.animatedValue as? Float ?: 0f

        for (i in 0 until rippleAmount) {
            // Stagger each ripple correctly
            var rippleProgress = progress - (i.toFloat() / rippleAmount)
            if (rippleProgress < 0) rippleProgress += 1.0f
            
            // Apply the smooth curve to the staggered progress
            val smoothProgress = interpolator.getInterpolation(rippleProgress)
            
            // Calculate transparency and scale based on the smooth progress
            paint.alpha = ((1f - smoothProgress) * 255).toInt()
            val currentRadius = rippleRadius * (1f + smoothProgress * (rippleScale - 1f))
            
            canvas.drawCircle(centerX, centerY, currentRadius, paint)
        }
    }

    fun startRippleAnimation() {
        if (!isRippleAnimationRunning) {
            isRippleAnimationRunning = true
            animator?.start()
        }
    }

    fun stopRippleAnimation() {
        if (isRippleAnimationRunning) {
            isRippleAnimationRunning = false
            animator?.cancel()
            postInvalidateOnAnimation()
        }
    }
}
